### PR TITLE
Attempt to backport the revamp of the communication engine

### PR DESCRIPTION
**Pull request by devreal**
My attempt at porting back the revamp of the communication infrastructure by @reazulhoque to current master. My goal was to condense the diff so I renamed `parsec_remote_dep.c` back to `remote_dep_mpi.c` and reshuffled things as much as possible to match current master. We can later rename files to what we think is appropriate.

At the moment, everything compiles (yay!) but the following tests fail locally (boo!):
- [x] 42/97 Test 42: collections/redistribute:mp .....................................***Failed    0.95 sec
- [x] 43/97 Test 43: collections/redistribute_random:mp ..............................***Failed    1.06 sec
- [x] 44/97 Test 44: collections/reshape .............................................***Failed    0.09 sec
- [x] 45/97 Test 45: collections/reshape:mt ..........................................***Exception: SegFault  0.33 sec
- [x] 46/97 Test 46: collections/reshape:mp ..........................................***Failed    1.85 sec
- [x] 47/97 Test 47: collections/reshape:mp:mt .......................................***Timeout 120.16 sec
- [x] 49/97 Test 49: collections/reshape/input_single_copy:mt ........................***Exception: SegFault  0.26 sec
- [x] 51/97 Test 51: collections/reshape/input_single_copy:mp:mt .....................***Failed    1.42 sec
- [x] 55/97 Test 55: collections/reshape/remote_multiple_outs_same_pred_flow:mp:mt ...***Failed    0.41 sec
- [x] 86/97 Test 86: dsl/dtd/pingpong:mp .............................................***Failed    0.51 sec
- [x] 90/97 Test 90: dsl/dtd/interleave_actions:mp ...................................***Failed    2.17 sec
- [x] 91/97 Test 91: dsl/dtd/allreduce:mp ............................................***Failed    0.21 sec
- [x] 97/97 Test 97: stencil_1D_mpi ..................................................***Timeout 120.26 sec

I'm hoping to crowdsource the debugging as much as possible, so feel free to give it a spin and push fixes. My guess is that many failures have the same root cause. People more knowledgeable in reshape and dtd may have an easier time identifying the issues...

The PR does *not* include the original history. I will put a reference to the original branch and attribute @reazulhoque but at this point it is futile trying to preserve it. The current history is a mess but once everything is in order we should squash before merging.  

I noticed that the CE API does not include the counterpart of `MPI_Pack_size`. Should we add it?

### CODE REVIEW

File: tests/dsl/dtd/dtd_test_ce.c [hunk 1]
```diff
@@ -0,0 +1,683 @@
...
+
+    if(ce->capabilites.supports_noncontiguous_datatype) {
+        parsec_datatype_t *datatype = malloc(sizeof(parsec_datatype_t));
+        parsec_type_create_contiguous(GET_activation_message->buf_size, parsec_datatype_int_t, datatype);
+        ce->mem_register(receive_buf, PARSEC_MEM_TYPE_NONCONTIGUOUS,
+                         1, *datatype,
+                         -1,
+                         &rank_1_memory_handle, &rank_1_memory_handle_size);
+    } else {
+         ce->mem_register(receive_buf, PARSEC_MEM_TYPE_CONTIGUOUS,
```
Comment (therault) [inline tests/dsl/dtd/dtd_test_ce.c:147]: Reading the code in `parsec_mpi_funnelled.c`, I don't see any places where the datatype can be -1. Some compilers complain that we cast an int (-1) into a pointer, which is right (to complain). Did I miss where `(void*)-1` is a special datatype constant? Should we define it (this issue occurs 4 times in this file, I will add a comment for each.

File: tests/dsl/dtd/dtd_test_ce.c [hunk 2]
```diff
@@ -0,0 +1,683 @@
...
+    if(ce->capabilites.supports_noncontiguous_datatype) {
+        parsec_datatype_t *datatype = malloc(sizeof(parsec_datatype_t));
+        parsec_type_create_contiguous(PUT_activation_message->buf_size, parsec_datatype_int_t, datatype);
+        ce->mem_register(receive_buf, PARSEC_MEM_TYPE_NONCONTIGUOUS,
+                         1, *datatype,
+                         -1,
+                         &rank_1_memory_handle, &rank_1_memory_handle_size);
+    } else {
+         ce->mem_register(receive_buf, PARSEC_MEM_TYPE_CONTIGUOUS,
+                          -1, -1,
```
Comment (therault) [inline tests/dsl/dtd/dtd_test_ce.c:281]: See comment above about -1 is not a valid datatype

File: tests/dsl/dtd/dtd_test_ce.c [hunk 3]
```diff
@@ -0,0 +1,683 @@
...
+            if(ce->capabilites.supports_noncontiguous_datatype) {
+                parsec_datatype_t *datatype = malloc(sizeof(parsec_datatype_t));
+                parsec_type_create_contiguous(buf_size, parsec_datatype_int_t, datatype);
+                ce->mem_register(send_buf, PARSEC_MEM_TYPE_NONCONTIGUOUS,
+                                 1, *datatype,
+                                 -1,
+                                 &rank_0_memory_handle, &rank_0_memory_handle_size);
+            } else {
+                 ce->mem_register(send_buf, PARSEC_MEM_TYPE_CONTIGUOUS,
+                                  -1, -1,
```
Comment (therault) [inline tests/dsl/dtd/dtd_test_ce.c:549]: See comment above about -1 not being a valid datatype

File: tests/dsl/dtd/dtd_test_ce.c [hunk 4]
```diff
@@ -0,0 +1,683 @@
...
```
Comment (therault) [inline tests/dsl/dtd/dtd_test_ce.c:623]: See comment above about -1 not being a valid datatype

File: parsec/remote_dep.h [hunk 5]
```diff
@@ -29,33 +38,30 @@ typedef ptrdiff_t remote_dep_datakey_t;
 #define PARSEC_ACTION_RESHAPE_REMOTE_ON_RELEASE  0x80000000
 #define PARSEC_ACTION_RELEASE_REMOTE_DEPS        (PARSEC_ACTION_SEND_INIT_REMOTE_DEPS | PARSEC_ACTION_SEND_REMOTE_DEPS)
 
-typedef struct remote_dep_wire_activate_s
-{
+enum {
```
Comment (therault) [inline parsec/remote_dep.h:41]: Missing `typedef` in front of `enum`. Surprisingly, this does not create an error unless you try to link the Fortran files (in which case the `parsec_remote_dep_tag_t` /variable/ is multiply-defined).

File: tests/dsl/dtd/dtd_test_ce.c [hunk 6]
```diff
@@ -0,0 +1,683 @@
...
+int
+callback_tag_2(parsec_comm_engine_t *ce,
+               parsec_ce_tag_t tag,
+               void *msg,
+               size_t msg_size,
+               int src,
+               void *cb_data)
+{
+    (void) ce; (void) cb_data;
+    printf("[%d] In callback for tag %llu, message sent from %d size: %ld message: ", my_rank, tag, src, msg_size);
```
Comment (therault) [inline tests/dsl/dtd/dtd_test_ce.c:58]: Replace `%llu` with `%"PRIu64"

File: tests/dsl/dtd/dtd_test_ce.c [hunk 7]
```diff
@@ -0,0 +1,683 @@
...
+int
+callback_tag_3(parsec_comm_engine_t *ce,
+               parsec_ce_tag_t tag,
+               void *msg,
+               size_t msg_size,
+               int src,
+               void *cb_data)
+{
+    (void) ce; (void) cb_data;
+    printf("[%d] In callback for tag %llu, message sent from %d size: %ld message: ", my_rank, tag, src, msg_size);
```
Comment (therault) [inline tests/dsl/dtd/dtd_test_ce.c:87]: Replace `%llu` with `%"PRIu64"

**Review by therault — CHANGES_REQUESTED** [review]

File: tests/dsl/dtd/dtd_test_ce.c (hunk 6 above)
Comment (devreal) [reply tests/dsl/dtd/dtd_test_ce.c:58]: Argh, those were debug statement that shouldn't have made it into the commit. We'll remove them

File: parsec/remote_dep_mpi.c [hunk 8]
```diff
@@ -1399,344 +1309,21 @@ static void remote_dep_mpi_profiling_init(void)
...
-    dep_comm = MPI_COMM_NULL;
+#endif  /* PARSEC_PROF_TRACE */
 
-    (void)context;
-    return 0;
-}
 
 static int remote_dep_mpi_on(parsec_context_t* context)
 {
-    remote_dep_mpi_setup(context);
```
[the remark this answers is not in the document]
Comment (devreal) [reply parsec/remote_dep_mpi.c:1739]: As far as I can see, the code in `remote_dep_mpi_setup` was moved to the MPI backend

File: tests/dsl/dtd/dtd_test_ce.c (hunk 1 above)
Comment (omor1) [reply tests/dsl/dtd/dtd_test_ce.c:147]: `PARSEC_DATATYPE_NULL` should be used, not `(void*)(-1)`. This was supposed to separate out cases where the comm engine backend supports non-contiguous datatypes (MPI) and that don't (LCI), but splitting up the logic in `mem_register` isn't really necessary—it's important that the comm engine exposes its capabilities, but computing the size of data as size of datatype * count can be done inside of `mem_register`.

Comment (omor1) [reply tests/dsl/dtd/dtd_test_ce.c:147]: Oh I see now you've already fixed that. :)

**Comment by therault** [comment]
I did some DPLASMA runs on Summit. Smallish size, POTRF, GEMM and GEQRF, with GPUs and without depending on the case. Used a tile size of 128x128 for CPUs and 1024x1024 for GPUs. Up to 64 nodes (128 processes), but some results are still pending.

Here is what I observe at this time. Sometimes, we get the revamp result and not the master result, but it was because the script was interrupted before the missing point could run.

summit-cpu
summit-1gpu
summit-2gpu
summit-3gpu

File: parsec/remote_dep_mpi.c [hunk 9]
```diff
@@ -2152,67 +1692,105 @@ remote_dep_mpi_put_start(parsec_execution_stream_t* es,
...
+                                   -1, NULL, // TODO JS: this interface is so broken, fix it!
+                                   dtt_size, // TODO JS: what about nbdtt? Is it ok to ignore it?!
+                                   &source_memory_handle, &source_memory_handle_size);
+
+        }
+
+        parsec_ce_mem_reg_handle_t remote_memory_handle = item->cmd.activate.remote_memory_handle;
+
 #if defined(PARSEC_DEBUG_NOISIER)
         MPI_Type_get_name(dtt, type_name, &len);
```
Comment (devreal) [inline parsec/remote_dep_mpi.c:1726]: Do we have a parsec-level replacement for `MPI_Type_get_name`?

Comment (bosilca) [reply parsec/remote_dep_mpi.c:1726]: nope, but maybe we need one.

Comment (omor1) [reply parsec/remote_dep_mpi.c:1726]: I think in my branch it's just surrounded by `#if defined(PARSEC_HAVE_MPI)`.

**Pull request merged**